Python/parking_lot.c

```diff
@@ -0,0 +1,470 @@
+    }
+#elif defined(USE_SEMAPHORES)
+    if (sem_init(&sema->platform_sem, /*pshared=*/0, /*value=*/0) < 0) {
+        Py_FatalError("parking_lot: sem_init failed");
+    }
+#else
+    if (pthread_mutex_init(&sema->mutex, NULL) != 0) {
+        Py_FatalError("parking_lot: pthread_mutex_init failed");
+    }
+    if (pthread_cond_init(&sema->cond, NULL)) {
+        Py_FatalError("parking_lot: pthread_cond_init failed");
+    }
+#endif
+}
+
+static void
+_PySemaphore_Destroy(_PySemaphore *sema)
+{
+#if defined(MS_WINDOWS)
+    CloseHandle(sema->platform_sem);
+#elif defined(USE_SEMAPHORES)
+    sem_destroy(&sema->platform_sem);
+#else
+    pthread_mutex_destroy(&sema->mutex);
+    pthread_cond_destroy(&sema->cond);
+#endif
+}
+
+static int
+_PySemaphore_PlatformWait(_PySemaphore *sema, _PyTime_t timeout)
+{
+    int res = Py_PARK_INTR;
+#if defined(MS_WINDOWS)
+    DWORD wait;
+    DWORD millis = 0;
+    if (timeout < 0) {
+        millis = INFINITE;
+    }
+    else {
+        millis = (DWORD) (timeout / 1000000);
+    }
+    wait = WaitForSingleObjectEx(sema->platform_sem, millis, FALSE);
+    if (wait == WAIT_OBJECT_0) {
+        res = Py_PARK_OK;
+    }
+    else if (wait == WAIT_TIMEOUT) {
+        res = Py_PARK_TIMEOUT;
+    }
+#elif defined(USE_SEMAPHORES)
+    int err;
+    if (timeout >= 0) {
+        struct timespec ts;
+
+        _PyTime_t deadline = _PyTime_Add(_PyTime_GetSystemClock(), timeout);
+        _PyTime_AsTimespec(deadline, &ts);
+
+        err = sem_timedwait(&sema->platform_sem, &ts);
+    }
+    else {
+        err = sem_wait(&sema->platform_sem);
+    }
+    if (err == -1) {
+        err = errno;
+        if (err == EINTR) {
+            res = Py_PARK_INTR;
+        }
+        else if (err == ETIMEDOUT) {
+            res = Py_PARK_TIMEOUT;
+        }
+        else {
+            _Py_FatalErrorFormat(__func__,
+                "unexpected error from semaphore: %d",
+                err);
+        }
+    }
+    else {
+        res = Py_PARK_OK;
+    }
+#else
+    pthread_mutex_lock(&sema->mutex);
+    if (sema->counter == 0) {
+        int err;
+        if (timeout >= 0) {
+            struct timespec ts;
+
+            _PyTime_t deadline = _PyTime_Add(_PyTime_GetSystemClock(), timeout);
+            _PyTime_AsTimespec(deadline, &ts);
+
+            err = pthread_cond_timedwait(&sema->cond, &sema->mutex, &ts);
+        }
+        else {
+            err = pthread_cond_wait(&sema->cond, &sema->mutex);
+        }
+        if (err) {
+            res = Py_PARK_TIMEOUT;
+        }
+    }
+    if (sema->counter > 0) {
+        sema->counter--;
+        res = Py_PARK_OK;
+    }
+    pthread_mutex_unlock(&sema->mutex);
+#endif
+    return res;
+}
+
+int
+_PySemaphore_Wait(_PySemaphore *sema, _PyTime_t timeout, int detach)
+{
+    PyThreadState *tstate = _PyThreadState_GET();
+    int was_attached = 0;
+    if (tstate) {
+        was_attached = (tstate->_status.active);
```

Comment:
We should avoid coupling this to the implementation details of `PyThreadState` (or of the GIL).  Additionally, `tstate->_status.active` does not imply the GIL is held.  Rather, it implies that `_PyThreadState_GET()` would return that tstate, which we already know because it came from there. 😄

It would be better if we used a dedicated function for this:

```suggestion
        was_attached = _PyEval_HoldsLock(tstate);
```

The implementation would go in ceval_gil.c and look something like:

```C
int
_PyEval_HoldsLock(PyThreadState *tstate)
{
    if (((PyThreadState*)_Py_atomic_load_relaxed(&gil->last_holder)) != tstate) {
        return 0;
    }
    return _Py_atomic_load_relaxed(&gil->locked);
}
```

We have a `holds_gil()` in Python/pystate.c, but I'm pretty sure it isn't adequate for our needs here.

Reply:
Ooof, this is tricky. I think the proposed `_PyEval_HoldsLock` has a race condition: 

1) We may be the `last_holder`, but with the GIL unlocked
2) By the time we read `&gil->locked` some other thread may have acquired the GIL so `gil->locked` is 1.

i.e. we may incorrectly think we hold the GIL when some other thread holds it.

I've changed it to just check that there is a non-NULL thread-state. Basically, if `detached` is 1 and there is a non-NULL thread state, then it's now the caller's responsibility to ensure that it holds the GIL.

We'll want to revisit this anyways as further pieces of PEP 703 get integrated, because `--disable-gil` builds will need something like https://peps.python.org/pep-0703/#thread-states.

Reply:
> I think the proposed _PyEval_HoldsLock has a race condition:

It works if we flip it around, doesn't it?

```C
int
_PyEval_HoldsLock(PyThreadState *tstate)
{
    if (!_Py_atomic_load_relaxed(&gil->locked)) {
        return 0;
    }
    return tstate == ((PyThreadState*)_Py_atomic_load_relaxed(&gil->last_holder));
}
```

Reply:
> if detached is 1 and there is a non-NULL thread state, then it's now the caller's responsibility to ensure that it holds the GIL.

Yeah, that's a pretty common pattern.  Many C-API functions require that the GIL be held.

> We'll want to revisit this anyways as further pieces of PEP 703 get integrated, because --disable-gil builds will need something like https://peps.python.org/pep-0703/#thread-states.

+1